Fix detail.php + MailWatch.pm for "Report:" from MailScanner

stefaweb opened the pull request:
- Fix wrong special characters parsing of $spamreport, $mcpreport, $text, $subject and $report in MailWatch.pm.
- Fixed space in detail.php
- Fixed 3 missing localization variables in edit.php
- Added fix_latin() for $headers in MailWatch.pm
- Comments cleanup in MailWatch.pm

stefaweb commented:
I still have a problem with the last "MailScanner:" without <br>.

![sc](https://cloud.githubusercontent.com/assets/12907102/23100680/60bccb1a-f687-11e6-8b3c-7409a816b8ba.png)

Skywalker-11 commented:
Can't you insert a br whereever  `Mailscanner:` is?

stefaweb commented:
> Can't you insert a br whereever Mailscanner: is?

The curent system is not good. It look for comma. It will be better to look for `MailScanner:` and remove/insert the right stuff.

stefaweb commented:
But is it alway` MailScanner:` in `Report:`?

stefaweb commented:
Another example.

![sc2](https://cloud.githubusercontent.com/assets/12907102/23100945/6332030c-f68b-11e6-8147-751d84555d7d.png)

Original text in database:

```
MailScanner: A long gap in a name is often used to hide part of it (2) ,MailScanner: A long gap in a name is often used to hide part of it (2) MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js)
,MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js)
```

stefaweb commented:
Problem with line break solved in a4dc89a.

Skywalker-11 commented:
@stefaweb does it resolve all problems you mentioned?

stefaweb commented:
@Skywalker-11 

Not all, I still have the problem with `MailScanner:` without comma. The rest is ok.

Skywalker-11 commented:
Can it be that the comma is intended to show that the two reports are for the same attachment and if no comma is there it is for an other attachment?

Skywalker-11 commented:
This would place a line break before each `MailScanner:` and also replaces a comma if it exists
`preg_replace("/[,]?MailScanner:/", "<br>MailScanner:", $input_lines);`

stefaweb commented:
Seems to have only one attachment but not sure. Look the file from one of the example.

[Delivery-Details.zip](https://github.com/mailwatch/1.2.0/files/785732/Delivery-Details.zip)

We can may be replace the search of comma with a regex to search `MailScanner:` and remove comma and space before? We will see in practice if someone else of `MailScanner:` comme there!

stefaweb commented:
I got this with:  
`$row[$f] = nl2br(preg_replace("/[,]?MailScanner:/", "<br>MailScanner:", htmlentities($row[$f])));`

![capture d ecran 2017-02-19 a 11 33 27](https://cloud.githubusercontent.com/assets/12907102/23101591/6048a87e-f697-11e6-9353-4a3ffbdcad0d.png)

Original text:

```
MailScanner: A long gap in a name is often used to hide part of it (3) MailScanner: A long gap in a name is often used to hide part of it (2)
MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js)
,MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js) ,MailScanner: A long gap in a name is often used to hide part of it (2) ,MailScanner: A long gap in a name is often used to hide part of it (3)
```

Seems to have double in original text (and result).

Skywalker-11 commented:
This would be better because it prevents br at the start `preg_replace("/(.)[[:space:]]?[,]?MailScanner:/", "${1}<br>MailScanner:", $input_lines);`

You don't need the nl2br then.

stefaweb commented:
Doesn't work. Got this with:

```
        if ($fieldn === 'Report:') {
            $fieldn = __('report04');
            $row[$f] = preg_replace('/[^[:print:]]/', '', $row[$f]);
            preg_replace("/(.)[[:space:]]?[,]?MailScanner:/", "${1}<br>MailScanner:", htmlentities($row[$f]));
            $row[$f] = preg_replace("/<br \/>/", '<br>', $row[$f]);
        }
```


![capture d ecran 2017-02-19 a 11 47 35](https://cloud.githubusercontent.com/assets/12907102/23101670/5a72ab14-f699-11e6-9327-1416ebc93919.png)

stefaweb commented:
Works. I've forgotten `$row[$f] =` at start...

stefaweb commented:
In fact no completly.

I got this on screen:

`Notice: Undefined variable: 1 in /var/www/html/mailscanner/detail.php on line 208`

Skywalker-11 commented:
Forgot a backslash. This is how it works
`  if ($fieldn === 'Report:') {
            $fieldn = __('report04');
            $row[$f] = preg_replace("/(.)[[:space:]]?[,]?MailScanner:/", "\${1}<br>MailScanner:", $row[$f]);
        }
`

stefaweb commented:
Fine.

![capture d ecran 2017-02-19 a 12 34 23](https://cloud.githubusercontent.com/assets/12907102/23101948/e70d613a-f69f-11e6-8008-0fed52bf9836.png)

I need localization for de.php  ->  report04

We just have this problem, we seems to have double but the problem is may be located in MailWatch.pm.

```
MailScanner: A long gap in a name is often used to hide part of it (3) 
MailScanner: A long gap in a name is often used to hide part of it (2)
MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js)
MailScanner: JScript Scripts are dangerous in email (1Delivery-Details.js)
MailScanner: A long gap in a name is often used to hide part of it (2)
MailScanner: A long gap in a name is often used to hide part of it (3)
```

stefaweb commented:
The "double" problem is also in the original 1.0.3 code of MailWatch.

![capture d ecran 2017-02-19 a 12 43 54](https://cloud.githubusercontent.com/assets/12907102/23102008/3b8915aa-f6a1-11e6-92ec-9747f07a91ca.png)

asuweb commented:
If you're going to fix this for cosmetic purposes, then the line break should be on the comma (,) not on "MailScanner".

Here are some real world examples that need to be catered for if indeed this needs fixed.

```Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./7D50E61517.A4B32/13014_ZIP.zip MailScanner: JScript Scripts are dangerous in email (23014.js) ,Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./7D50E61517.A4B32/2179152343083.zip/3014_ZIP.zip ,MailScanner: JScript Scripts are dangerous in email (23014.js)```

```MailScanner: JScript Scripts are dangerous in email (231836.js) ,F-Secure: ./7C35C621EA.A19EC/131836_ZIP.zip: Infected: Trojan:JS/Kavala.D [FSE] Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./7C35C621EA.A19EC/131836_ZIP.zip MailScanner: JScript Scripts are dangerous in email (231836.js) ,F-Secure: [./7C35C621EA.A19EC/00145434151.zip] 31836_ZIP.zip: Infected: Trojan:JS/Kavala.D [FSE] Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./7C35C621EA.A19EC/00145434151.zip/31836_ZIP.zip```

```F-Secure: ./78563633B3.ADFCA/117961_ZIP.zip: Infected: Trojan:JS/Kavala.D [FSE] F-Secure: [./78563633B3.ADFCA/117961_ZIP.zip] 17961.js: Infected: Trojan.JS.Downloader.HOJ [Aquarius] Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./78563633B3.ADFCA/117961_ZIP.zip F-Secure: ./78563633B3.ADFCA/217961.js: Infected: Trojan.JS.Downloader.HOJ [Aquarius] MailScanner: JScript Scripts are dangerous in email (217961.js) ,F-Secure: [./78563633B3.ADFCA/10570.zip] 17961_ZIP.zip: Infected: Trojan:JS/Kavala.D [FSE] F-Secure: [./78563633B3.ADFCA/10570.zip] [17961_ZIP.zip] 17961.js: Infected: Trojan.JS.Downloader.HOJ [Aquarius] Sophos: >>> Virus 'Mal/DrodZp-A' found in file ./78563633B3.ADFCA/10570.zip/17961_ZIP.zip ,F-Secure: ./78563633B3.ADFCA/217961.js: Infected: Trojan.JS.Downloader.HOJ [Aquarius] MailScanner: JScript Scripts are dangerous in email (217961.js)```

stefaweb commented:
@asuweb: Seems the problem start in MailWatch.pm. Look at #557.

You example show that we can have more than `MailScanner:` in field `report`  (Infected:). 

We need to fix MailWatch.pm first.

stefaweb commented:
I added two missing localization variables and reverted back the test to develop.

Is normal to do not have the SpamAssassin detail when we have a dangerous content?

![capture d ecran 2017-02-19 a 16 28 26](https://cloud.githubusercontent.com/assets/12907102/23103747/b142013e-f6c0-11e6-84a4-5d0a9f91fa65.png)


![capture d ecran 2017-02-19 a 16 28 42](https://cloud.githubusercontent.com/assets/12907102/23103748/b16bae94-f6c0-11e6-9799-1431ef2f1248.png)

stefaweb commented:
Loaded a new version of MailWatch.pm.

Not sure if when we use [[:print:]], we still need to filter \n and \t. The perl doc is not clear on this (for me).

I sert a commented uniq function if needed to try.

stefaweb commented:
`s/[[:print:]]+//g` do not work. I got empty variable.

stefaweb commented:
Does someone know why no "g" at the end?

$text =~ s/\n/ /;  # Make sure text report only contains 1 line
$text =~ s/\t/ /; # and no tab characters

stefaweb commented:
develop version do not filter Line Feed (0a) for $text.

![sc](https://cloud.githubusercontent.com/assets/12907102/23122864/54307614-f766-11e6-947b-33ee0d8e271f.jpg)

stefaweb commented:
Hello,

Last uploaded version of MailWatch.pm seems to work fine. I tried with custom made spam email with .js attachment. No more LF or space (and may be double but not sure).

The uploaded version have debug on.

If someone can test it with more example.

![capture d ecran 2017-02-20 a 15 12 07](https://cloud.githubusercontent.com/assets/12907102/23128384/0bdd1534-f77f-11e6-929c-81ad8fc512ae.png)

stefaweb commented:
Ok, ready to merge.

stefaweb commented on MailScanner_perl_scripts/MailWatch.pm:
```diff
@@ -42,6 +42,9 @@ use POSIX;
 use Encoding::FixLatin qw(fix_latin);
 
+# Uncommet the folloging line when debugging SQLBlackWhiteList.pm
+use Data::Dumper;
```
done
